sherpa/utils/parallel.py

```diff
@@ -0,0 +1,462 @@
+
+    procs = [multiprocessing.Process(target=worker,
+                                     args=(function, ii, chunk, out_q, err_q))
+             for ii, chunk in enumerate(sequence)]
+
+    return run_tasks(procs, err_q, out_q)
+
+
+# TODO: this routine needs a review
+def parallel_map_funcs(funcs, datasets, numcores=None):
+    """Run a sequence of function on a sequence of inputs in parallel.
+
+    Sherpa's parallel_map runs a single function to an iterable set of
+    sequence.  parallel_map_funcs is generalized parallelized version
+    of sherpa's parallel_map function since each element of the ordered
+    iterable funcs shall operate on the each element of the datasets.
+
+    Parameters
+    ----------
+    funcs : a list or tuple of functions
+       An ordered iteratble sequence of functions which accepts an element
+       of the datasets and returns a value.  The number of elements in
+       funcs must match the number of elements of the datasets.
+    datasets : a list or tuple of array_like
+       The data to be passed to ``func``. The number of elements in
+       datasets must match the number of elements of funcs.
+    numcores : int or None, optional
+       The number of calls to ``funcs`` to run in parallel. When
+       set to ``None``, all the available CPUs on the machine - as
+       set either by the 'numcores' setting of the 'parallel' section
+       of Sherpa's preferences or by multiprocessing.cpu_count - are
+       used.
+
+    Returns
+    -------
+    ans : array
+       The return values from the calls, in the same order as the
+       ``sequence`` array.
+
+    Notes
+    -----
+    Due to the overhead involved in passing the functions and datasets
+    to the different cores, the functions should be very time consuming
+    to compute (of order 0.1-1s).  This is similar to the ``parallel_map``
+    function.
+
+    An ordered iterable (i.e. tuple or list) should be used to pass multiple
+    values to the multiple functions. The lengths of the iterable funcs and
+    datasets must be equal. The corresponding funcs and datasets are passed
+    to the different cores to distribute the work in parallel. There is no
+    guarantee to the ordering of the tasks.
+
+    Examples
+    --------
+
+    In the following examples a simple set of computations, sum and std
+    deviations, are used; in reality the function is expected to be run
+    on computations that take a significant amount of time to run.
+
+    Run the computation (summing up each element of the first input array
+    and calculate the standard deviation of the second input array)
+    on a separate core and return the results (unless the machine only
+    has a single core or the parallel.numcores setting is set to 1).
+
+    >>> import numpy as np
+    >>> funcs = [np.sum, np.std]
+    >>> datasets = [np.arange(3), np.arange(4)]
+    >>> parallel_map_funcs(funcs, datasets, numcores=2)
+    [0, 1, 2, 0.0, 0.0, 0.0, 0.0]
```

Comment:
Now I get 7 outputs, where I would have expected 2 (one per function). So apparently I didn ot understand who this works.

I can make a suggestion for re-phrasing the expalantion above, but I'm not sure I understand  the purpose.

Reply:
I really don't want to touch `parallel_map_funcs` or it's documentation (beyond making it work with doctestplus) because I do not understand what it is meant to be doing (this time I already have an issue about it - see #1743)

Reply:
I was skeptical about the parallel stuff at first and @DougBurke told me not to be afraid. I spend some time reading and I think I have better grasp now, at least to what this function does and I'm comfortable that is does what I wrote in my suggestion.
If you don't want this committed under you name, I can make a PR with just this documentation change when this one is merged and then you'll have to review my text...

Reply:
I'd rather this be done in a separate PR. I wonder if we even need the method at all as my guess it's part of an old attempt to speed things up that never went anywhere, but that would be better discussed in a separate PR ...

Reply:
I think I've actually used this method in a no-sherpa project, but I'll check. Either way, I can be addressed in a followup. See #1787